Add support for Markdown inline links

bdach opened the pull request:
# Summary

While reviewing #6542 it became apparent that there was another Markdown link format variant, used in comments that came from the web API, called the "inline link" style. It allows to specify the tooltip title within the actual URL portion, as such:
```
[link text](https://osu.ppy.sh "tooltip text")
```
Extend the Markdown parsing regex to allow parsing these inline links. Within the parenthesis () part of the Markdown URL syntax, introduce a new capturing group:
```
    (
      \s+              // whitespace between actual URL and inline title
      (?<title>        // start of "title" named group
        ""             // opening double quote (doubled inside @ string)
        (
          [^""]        // any character but a double quote
          |            // or
          (?<=\\)      // the next character should be preceded by a \
          ""           // a double quote
        )*             // zero or more times
        ""             // closing double quote
      )
    )?                 // the whole group is optional
```
This allows for parsing the inline links as-provided by web, with support for escaping double quotes inside of the title part. Correctness is displayed by the passing tests.

Testing with #6542 additionally surfaced a latent bug leading to a crash, caused by formatted links that had URLs in the display text, for example
```
[mean example - https://osu.ppy.sh](https://osu.ppy.sh)
```
In that case the outer Markdown link would get picked up once, and then reduced to the link text when looking for other links, leading to it being picked up again the second time when the raw link is found.

To resolve that case, add a check in the raw link parsing path that ensures that the found URL is not a part of a bigger, pre-existing link.

# Remarks

I'm not currently using the `title` named group to change mouseover text as web seems to do, since as discussed in Discord it could lead to users posting links with in chat or in comments with misleading tooltip text.

bdach committed:
Add extended testing for Markdown links

While reviewing #6542 it became apparent that there was another Markdown
link format variant, used in comments that came from the web API, called
the "inline link" style. It allows to specify the tooltip title within
the actual URL portion, as such:

    [link text](https://osu.ppy.sh "tooltip text")

Add tests with a couple of easy and trickier examples of such a format.
Moreover, add a new edge case of a Markdown link with a link inside
the display text, which during tests was detected to be problematic.

bdach committed:
Add support for parsing Markdown inline links

Extend the Markdown parsing regex to allow parsing so-called inline
links. Within the parenthesis () part of the Markdown URL syntax,
introduce a new capturing group:

    (
      \s+              // whitespace between actual URL and inline title
      (?<title>        // start of "title" named group
        ""             // opening double quote (doubled inside @ string)
        (
          [^""]        // any character but a double quote
          |            // or
          (?<=\\)      // the next character should be preceded by a \
          ""           // a double quote
        )*             // zero or more times
        ""             // closing double quote
      )
    )?                 // the whole group is optional

This allows for parsing the inline links as-provided by web. Correctness
is displayed by the passing tests.

bdach committed:
Resolve link-in-link edge case

Testing with #6542 surfaced a crash scenario, caused by formatted links
that had URLs in the display text, for example

    [mean example - https://osu.ppy.sh](https://osu.ppy.sh)

In that case the outer Markdown link would get picked up once, and then
reduced to the link text when looking for other links, leading to it
being picked up again the second time when the raw link is found.

Add a check in the raw link parsing path that ensures that the found
URL is not a part of a bigger, pre-existing link.

Zeroks77 commented on osu.Game.Tests/Chat/MessageFormatterTests.cs:
```diff
@@ -273,6 +273,66 @@ public void TestMarkdownFormatLinkWithEscapedAndBalancedBrackets()
+        [Test]
+        public void TestMarkdownFormatLinkWithInlineTitle()
+        {
+            Message result = MessageFormatter.FormatMessage(new Message { Content = "I haven't seen [this link format](https://osu.ppy.sh \"osu!\") before..." });
```
//Arragne 
var blub = "someStuff";
var expected = "SomeStuff";

//Act 
var result = doSomeStuff(blub)

//Assert 
Assert.Something(expected, result);
 ?


have a detailed look [here](https://docs.microsoft.com/en-us/visualstudio/test/unit-test-basics?view=vs-2019)

it would make the test more readable and mantainable

bdach replied:
I am familiar with the arrange-act-assert pattern (also known as the given-when-then pattern). I use it in my own projects. If you notice these tests also are arranged similarly, but don't have the comment lines - and I don't believe adding them is so important as to break existing conventions in this repository.,

Zeroks77 commented on osu.Game.Tests/Chat/MessageFormatterTests.cs:
```diff
@@ -273,6 +273,66 @@ public void TestMarkdownFormatLinkWithEscapedAndBalancedBrackets()
+        }
+
+        [Test]
+        public void TestMarkdownFormatLinkWithUrlInTextAndInlineTitle()
```
Also what does the test do? If it fails, if it works, please rename the test so you can see what it should do.

Example: ifAThanReturnSomething 

The test name must make clear what is tested and how

Zeroks77 replied:
same goes to every other Testname

bdach replied:
Again, I am familiar with that convention and sometimes use it, but choose not to here due to pre-existing conventions. The value of such naming is disputable to me, and anyone lacking context can check out the commit log and PR description.

Zeroks77 commented on osu.Game/Online/Chat/MessageFormatter.cs:
```diff
@@ -20,7 +20,7 @@ public static class MessageFormatter
         private static readonly Regex new_link_regex = new Regex(@"\[(?<url>[a-z]+://[^ ]+) (?<text>(((?<=\\)[\[\]])|[^\[\]])*(((?<open>\[)(((?<=\\)[\[\]])|[^\[\]])*)+((?<close-open>\])(((?<=\\)[\[\]])|[^\[\]])*)+)*(?(open)(?!)))\]");
 
         // [test](https://osu.ppy.sh/b/1234) -> test (https://osu.ppy.sh/b/1234) aka correct markdown format
```
Please provide a more detailed description of what the regex should do.

not everyone is able to understand regex right away.

bdach replied:
I'm not sure what I could explain further here without restating the regex verbatim, part-by-part, which seems counterproductive - this does not seem like the correct place to explain regular expression syntax. (And if you check the commit message, I explained the added fragment part-by-part - the explanation is in the PR description, too).

bdach committed:
Change containment check to overlap

Due to scenarios wherein a formatted link ended up as part of a larger
raw link after parsing, change the containment check to an overlap check
and add appropriate tests for these edge cases.

bdach commented:
I realised that the previous containment check was not strong enough and needs to be bumped up to an overlap test - changed and added appropriate test cases.

smoogipoo committed:
Merge branch 'master' into markdown-inline-links
